File: javascript/leet_code/463.mjs
/**
 * @fileoverview This module provides a function to calculate the perimeter of an island in a 2D grid.
 */

/**
 * Calculates the perimeter of an island in a 2D grid.
 * The grid is represented as a 2D array where 0 represents water and 1 represents land.
 * An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically.
 * Assume all four edges of the grid are surrounded by water.
 *
 * @param {number[][]} grid - The 2D grid representation of the map.
 * @return {number} The perimeter of the island.
 *
 * @throws {TypeError} If the input grid is not a 2D array.
 * @throws {Error} If the grid does not contain any land (1s).
 */
let islandPerimeter = function (grid) {
    // Validate the input grid.
    if (!Array.isArray(grid) || !Array.isArray(grid[0])) {
        throw new TypeError('The input grid must be a 2D array.');
    }

    let peri = 0;  // Initialize the perimeter of the island.

    // Iterate over the rows of the grid.
    for (let i = 0; i < grid.length; i++) {
        // Iterate over the columns of the grid.
        for (let j = 0; j < grid[0].length; j++) {
            // If the current cell is land, calculate the possible perimeter.
            if (1 === grid[i][j]) {
                let poss = 4;  // Initialize the possible perimeter.

                // Check the left cell.
                if (1 === grid[i][j - 1]) {
                    poss--;
                }

                // Check the right cell.
                if (1 === grid[i][j + 1]) {
                    poss--;
                }

                // Check the bottom cell.
                if (i + 1 < grid.length) {
                    if (1 === grid[i + 1][j]) {
                        poss--;
                    }
                }

                // Check the top cell.
                if (0 <= i - 1) {
                    if (1 === grid[i - 1][j]) {
                        poss--;
                    }
                }

                // Add the possible perimeter to the total perimeter.
                peri += poss;
            }
        }
    }

    // If the perimeter is still 0, there is no land in the grid.
    if (0 === peri) {
        throw new Error('The grid does not contain any land.');
    }

    // Return the perimeter of the island.
    return peri;
};

File: ruby/leet_code/463.rb
# Function to calculate the perimeter of the islands in a grid
#
# This function calculates the perimeter of the islands in a 2D grid map of '1's (land) and '0's (water).
# An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically.
# The perimeter is calculated by counting the edges of the land cells.
#
# @param grid [Array<Array<Integer>>] The 2D grid map of '1's (land) and '0's (water)
# @return [Integer] The total perimeter of the islands
#
# @example
#   island_perimeter([[0,1,0,0],[1,1,1,0],[0,1,0,0],[1,1,0,0]]) #=> 16
#
def island_perimeter(grid)
  count = 0
  max_row_idx = grid.length - 1
  max_col_idx = grid.first.length - 1

  # Iterate over each cell in the grid
  (0..max_row_idx).each do |row|
    (0..max_col_idx).each do |col|
      # Skip if the cell is water
      next if grid[row][col] == 0

      # Start with 4 edges for a land cell
      edges = 4

      # If there is a land cell below the current cell, subtract 2 edges
      if row < max_row_idx
        edges -= 2 if grid[row + 1][col] == 1
      end

      # If there is a land cell to the right of the current cell, subtract 2 edges
      if col < max_col_idx
        edges -= 2 if grid[row][col + 1] == 1
      end

      # Add the remaining edges to the total count
      count += edges
    end
  end

  # Return the total count of edges, which is the perimeter of the islands
  count
end